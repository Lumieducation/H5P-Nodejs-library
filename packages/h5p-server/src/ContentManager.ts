import { Stream, Readable } from 'stream';

import { ContentMetadata } from './ContentMetadata';
import {
    ContentId,
    ContentParameters,
    IContentMetadata,
    IContentStorage,
    IContentUserDataStorage,
    IUser,
    Permission
} from './types';

import Logger from './helpers/Logger';

const log = new Logger('ContentManager');

/**
 * The ContentManager takes care of saving content and dependent files. It only contains storage-agnostic functionality and
 * depends on a ContentStorage object to do the actual persistence.
 */
export default class ContentManager {
    /**
     * @param contentStorage The storage object
     * @param contentUserDataStorage The contentUserDataStorage to delete contentUserData for content when it is deleted
     */
    constructor(
        public contentStorage: IContentStorage,
        public contentUserDataStorage?: IContentUserDataStorage
    ) {
        log.info('initialize');
    }

    /**
     * Adds a content file to an existing content object. The content object has to be created with createContent(...) first.
     * @param contentId The id of the content to add the file to
     * @param filename The name of the content file
     * @param stream A readable stream that contains the data
     * @param user The user who owns this object
     * @returns
     */
    public async addContentFile(
        contentId: ContentId,
        filename: string,
        stream: Stream,
        user: IUser
    ): Promise<void> {
        log.info(`adding file ${filename} to content ${contentId}`);
        return this.contentStorage.addFile(contentId, filename, stream, user);
    }

    /**
     * Checks if a piece of content exists.
     * @param contentId the content to check
     * @returns true if the piece of content exists
     */
    public async contentExists(contentId: ContentId): Promise<boolean> {
        log.debug(`checking if content ${contentId} exists`);
        return this.contentStorage.contentExists(contentId);
    }

    /**
     * Checks if a file exists.
     * @param contentId The id of the content to add the file to
     * @param filename the filename of the file to get
     * @returns true if the file exists
     */
    public contentFileExists = async (
        contentId: ContentId,
        filename: string
    ): Promise<boolean> => this.contentStorage.fileExists(contentId, filename);

    /**
     * Creates a content object in the repository. Add files to it later with addContentFile(...).
     * @param metadata The metadata of the content (= h5p.json)
     * @param content the content object (= content/content.json)
     * @param user The user who owns this object.
     * @param contentId (optional) The content id to use
     * @returns The newly assigned content id
     */
    public async createOrUpdateContent(
        metadata: IContentMetadata,
        content: ContentParameters,
        user: IUser,
        contentId?: ContentId
    ): Promise<ContentId> {
        log.info(`creating content for ${contentId}`);
        return this.contentStorage.addContent(
            metadata,
            content,
            user,
            contentId
        );
    }

    /**
     * Deletes a piece of content, the corresponding contentUserData and all files dependent on it.
     * @param contentId the piece of content to delete
     * @param user the user who wants to delete it
     */
    public async deleteContent(
        contentId: ContentId,
        user: IUser
    ): Promise<void> {
        await this.contentStorage.deleteContent(contentId, user);

        if (this.contentUserDataStorage) {
            try {
                await this.contentUserDataStorage.deleteAllContentUserDataByContentId(
                    contentId
                );
            } catch (error) {
                log.error(
                    `Could not delete content user data with contentId ${contentId}`
                );
                log.error(error);
            }
            try {
                await this.contentUserDataStorage.deleteFinishedDataByContentId(
                    contentId
                );
            } catch (error) {
                log.error(
                    `Could not finished data with contentId ${contentId}`
                );
                log.error(error);
            }
        }
    }

    /**
     * Deletes a file from a content object.
     * @param contentId the content object the file is attached to
     * @param filename the file to delete
     */
    public async deleteContentFile(
        contentId: ContentId,
        filename: string
    ): Promise<void> {
        return this.contentStorage.deleteFile(contentId, filename);
    }

    /**
     * Returns a readable stream of a content file (e.g. image or video) inside a piece of content
     * @param contentId the id of the content object that the file is attached to
     * @param filename the filename of the file to get
     * @param user the user who wants to retrieve the content file
     * @param rangeStart (optional) the position in bytes at which the stream should start
     * @param rangeEnd (optional) the position in bytes at which the stream should end
     * @returns
     */
    public async getContentFileStream(
        contentId: ContentId,
        filename: string,
        user: IUser,
        rangeStart?: number,
        rangeEnd?: number
    ): Promise<Readable> {
        log.debug(`loading ${filename} for ${contentId}`);

        return this.contentStorage.getFileStream(
            contentId,
            filename,
            user,
            rangeStart,
            rangeEnd
        );
    }

    /**
     * Returns the metadata (=contents of h5p.json) of a piece of content.
     * @param contentId the content id
     * @param user The user who wants to access the content
     * @returns
     */
    public async getContentMetadata(
        contentId: ContentId,
        user: IUser
    ): Promise<IContentMetadata> {
        // We don't directly return the h5p.json file content as
        // we have to make sure it conforms to the schema.
        return new ContentMetadata(
            await this.contentStorage.getMetadata(contentId, user)
        );
    }

    /**
     * Returns the content object (=contents of content/content.json) of a piece of content.
     * @param contentId the content id
     * @param user The user who wants to access the content
     * @returns
     */
    public async getContentParameters(
        contentId: ContentId,
        user: IUser
    ): Promise<ContentParameters> {
        return this.contentStorage.getParameters(contentId, user);
    }

    /**
     * Returns an array of permissions a user has on a piece of content.
     * @param contentId the content to check
     * @param user the user who wants to access the piece of content
     * @returns an array of permissions
     */
    public async getUserPermissions(
        contentId: ContentId,
        user: IUser
    ): Promise<Permission[]> {
        log.info(`checking user permissions for ${contentId}`);
        return this.contentStorage.getUserPermissions(contentId, user);
    }

    /**
     * Lists the content objects in the system (if no user is specified) or owned by the user.
     * @param user (optional) the user who owns the content
     * @returns a list of contentIds
     */
    public listContent(user?: IUser): Promise<ContentId[]> {
        return this.contentStorage.listContent(user);
    }

    /**
     * Gets the filenames of files added to the content with addContentFile(...) (e.g. images, videos or other files)
     * @param contentId the piece of content
     * @param user the user who wants to access the piece of content
     * @returns a list of files that are used in the piece of content, e.g. ['image1.png', 'video2.mp4']
     */
    public async listContentFiles(
        contentId: ContentId,
        user: IUser
    ): Promise<string[]> {
        log.info(`loading content files for ${contentId}`);
        return this.contentStorage.listFiles(contentId, user);
    }

    public sanitizeFilename = (filename: string): string => {
        if (this.contentStorage.sanitizeFilename) {
            return this.contentStorage.sanitizeFilename(filename);
        }
        return filename;
    };
}
