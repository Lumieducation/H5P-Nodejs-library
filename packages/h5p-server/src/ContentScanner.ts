import Logger from './helpers/Logger';
import LibraryManager from './LibraryManager';
import LibraryName from './LibraryName';
import { ILibraryName, ISemanticsEntry } from './types';

const log = new Logger('ContentScanner');

/**
 * @param semantics The semantic information about the current content params
 * @param params The current params
 * @param jsonPath The current path of the object in the tree as a JSON path
 * (example: .media.type.path)
 * @param parentParams The parent element in the params tree. Can be used to
 * delete this entry.
 * @returns true if the children of this entry should not be scanned (= abort
 * scan for this subtree)
 */
export type ScanCallback = (
    semantics: ISemanticsEntry,
    params: any,
    jsonPath: string,
    parentParams: any
) => boolean;

/**
 * Scans the content parameters (= content.json) of a piece of content and calls
 * a callback for each element in the semantic tree. This includes all nested
 * pieces of content.
 *
 * You must pass ins a callback that is called for every element in the params
 * tree. If this callback returns true, the scan will be aborted _for the
 * element and its children_, but not for the rest of the params.
 */
export class ContentScanner {
    constructor(private libraryManager: LibraryManager) {
        log.info('initialize');
    }

    /**
     * Scans the specified parameters and executes the callback for every
     * element in the tree. This includes nested content!
     * @param params the params to scan
     * @param mainLibraryName the library name
     * @param callback a function that is executed for every element in the
     * semantic structure. Return true in it to abort the scan for the element
     * itself and all of its children (the rest of the scan will continue).
     */
    public async scanContent(
        params: any,
        mainLibraryName: ILibraryName,
        callback: ScanCallback
    ): Promise<void> {
        log.debug(
            `scanning content for of type ${LibraryName.toUberName(
                mainLibraryName
            )}`
        );

        const mainSemantics =
            await this.libraryManager.getSemantics(mainLibraryName);
        await this.walkSemanticsRecursive(
            mainSemantics,
            params,
            '$',
            callback,
            {
                doNotAddNameToJsonPath: false
            }
        );
    }

    /**
     * Walks through an element in the semantic tree of a library.
     * @param elementSemantics the semantic information for the current element
     * @param elementParams the parameters for the current element (as in
     * content.json)
     * @param parentJsonPath the JSON path of the parent (example: .media.type)
     * @param parentParams the parent params in the params tree
     * @param callback a function that is executed for this element and for
     * every child
     * @param doNotAddNameToJsonPath if true, the name of the current element
     * will not appended to the JSON path This is the case when a group contains
     * only one element. Then the child's name is omitted in the parameters.
     */
    private async walkEntryRecursive(
        elementSemantics: ISemanticsEntry,
        elementParams: any,
        parentJsonPath: string,
        parentParams: any,
        callback: ScanCallback,
        options: {
            doNotAddNameToJsonPath: boolean;
        } = {
            doNotAddNameToJsonPath: false
        }
    ): Promise<void> {
        if (elementParams === undefined && !elementSemantics.optional) {
            log.info(
                `${elementSemantics.name} has no params but is not optional.`
            );
        }

        // we ignore elements that are not used in the parameters
        if (elementParams === undefined) {
            return;
        }

        // Don't append our name to the JSON path if our parent signalled us not
        // to do so
        const currentJsonPath = options.doNotAddNameToJsonPath
            ? parentJsonPath
            : `${parentJsonPath}.${elementSemantics.name}`;

        // Treating a special case of how group deals with fields with only one
        // entry.
        if (
            elementSemantics.type === 'group' &&
            elementSemantics.fields.length === 1 &&
            !elementParams[elementSemantics.name]
        ) {
            // The parameters produced by H5P are weird in this case: You would
            // expect the parameters to be an array with a single entry [{...}],
            // as the semantic structure defines a group with a single entry.
            // For some reason, H5P directly puts the object {...} into the
            // parameters. We have to compensate for this special case.

            log.debug(`found single group entry ${currentJsonPath}`);
            await this.walkEntryRecursive(
                { ...elementSemantics.fields[0], name: elementSemantics.name },
                elementParams,
                options.doNotAddNameToJsonPath
                    ? parentJsonPath
                    : `${parentJsonPath}.${elementSemantics.name}`,
                parentParams,
                callback,
                {
                    // We have already added the parent's name to the JSON path,
                    // so we don't want the child to add its name.
                    doNotAddNameToJsonPath: true
                }
            );
            return;
        }

        if (
            callback(
                elementSemantics,
                elementParams,
                currentJsonPath,
                parentParams
            )
        ) {
            // don't walk further into the tree if the callback signalled to stop
            return;
        }

        switch (elementSemantics.type) {
            case 'library': {
                // If an element contains another library, we have to retrieve
                // the exact name, and the nested content parameters.
                if (elementParams.library === undefined) {
                    // Skip if the element is empty. (= unused)
                    return;
                }
                const subLibraryName = LibraryName.fromUberName(
                    elementParams.library,
                    {
                        useWhitespace: true
                    }
                );
                const subSemantics =
                    await this.libraryManager.getSemantics(subLibraryName);
                await this.walkSemanticsRecursive(
                    subSemantics,
                    elementParams.params,
                    `${currentJsonPath}.params`,
                    callback,
                    { doNotAddNameToJsonPath: false }
                );
                break;
            }
            case 'group':
                // groups contain several semantic entries, each with their own
                // parameters.
                for (const groupElement of elementSemantics.fields) {
                    await this.walkEntryRecursive(
                        groupElement,
                        elementParams[groupElement.name],
                        currentJsonPath,
                        elementParams,
                        callback,
                        { doNotAddNameToJsonPath: false }
                    );
                }
                break;
            case 'list': {
                // lists contain one semantic entry, but several content
                // elements
                let counter = 0;
                for (const listElement of elementParams) {
                    await this.walkEntryRecursive(
                        elementSemantics.field,
                        listElement,
                        `${currentJsonPath}[${counter}]`,
                        elementParams,
                        callback,
                        {
                            // We don't want the field name of a list to be
                            // added to the JSON path.
                            doNotAddNameToJsonPath: true
                        }
                    );
                    counter += 1;
                }
                break;
            }
            default:
                break;
        }
    }

    /**
     * Walks through all semantic entries in a library semantics.
     * @param semantics the semantic structure of a library
     * @param params the parameter object for the content
     * @param parentJsonPath the path of the parent
     * @param callback the callback to execute for every element in the tree
     */
    private async walkSemanticsRecursive(
        semantics: ISemanticsEntry[],
        params: any,
        parentJsonPath: string,
        callback: ScanCallback,
        options: {
            doNotAddNameToJsonPath: boolean;
        } = {
            doNotAddNameToJsonPath: false
        }
    ): Promise<void> {
        for (const semanticEntry of semantics) {
            await this.walkEntryRecursive(
                semanticEntry,
                params[semanticEntry.name],
                parentJsonPath,
                params,
                callback,
                options
            );
        }
    }
}
