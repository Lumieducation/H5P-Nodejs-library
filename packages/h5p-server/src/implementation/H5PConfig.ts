import { IH5PConfig, IKeyValueStorage } from '../types';

/**
 * Stores configuration options and literals that are used throughout the
 * system. Also loads and saves the configuration of changeable values (only
 * those as "user-configurable") in the storage object.
 */
export default class H5PConfig implements IH5PConfig {
    /**
     * @param storage A key-value storage object that persists the changes to
     * the disk or gets them from the implementation/plugin
     * @param defaults default values to use instead of the ones set by this
     * class
     */
    constructor(storage?: IKeyValueStorage, defaults?: Partial<IH5PConfig>) {
        this.storage = storage;
        if (defaults) {
            for (const key in defaults) {
                if (this[key] !== undefined) {
                    this[key] = defaults[key];
                }
            }
        }
    }
    public ajaxUrl: string = '/ajax';
    public baseUrl: string = '/h5p';
    public contentFilesUrl: string = '/content';
    public contentFilesUrlPlayerOverride: string;
    public contentTypeCacheRefreshInterval: number = 1 * 1000 * 60 * 60 * 24;
    public contentHubEnabled: boolean = false;
    public contentHubMetadataRefreshInterval: number = 1 * 1000 * 60 * 60 * 24;
    public contentUserDataUrl: string = '/contentUserData';
    public contentWhitelist: string =
        'json png jpg jpeg gif bmp tif tiff svg eot ttf woff woff2 otf webm mp4 ogg mp3 m4a wav txt pdf rtf doc docx xls xlsx ppt pptx odt ods odp xml csv diff patch swf md textile vtt webvtt gltf glb';
    public coreApiVersion: { major: number; minor: number } = {
        major: 1,
        minor: 24
    };
    public coreUrl: string = '/core';
    public customization: {
        global: {
            editor?: {
                scripts?: string[];
                styles?: string[];
            };
            player?: {
                scripts?: string[];
                styles?: string[];
            };
        };
    } = {
        global: {
            editor: {
                scripts: [],
                styles: []
            },
            player: {
                scripts: [],
                styles: []
            }
        }
    };
    public disableFullscreen: boolean = false;
    public downloadUrl: string = '/download';
    public editorAddons?: {
        [machineName: string]: string[];
    };
    public editorLibraryUrl: string = '/editor';
    public enableLrsContentTypes: boolean = true;
    public exportMaxContentPathLength: number = 255;
    public fetchingDisabled: 0 | 1 = 0;
    public h5pVersion: string = '1.24-master';
    public hubContentTypesEndpoint: string =
        'https://api.h5p.org/v1/content-types/';
    public hubRegistrationEndpoint: string = 'https://api.h5p.org/v1/sites';
    public installLibraryLockMaxOccupationTime: number = 10000;
    public installLibraryLockTimeout: number = 20000;
    public contentHubContentEndpoint: string =
        'https://hub-api.h5p.org/v1/contents';
    public contentHubMetadataEndpoint: string =
        'https://hub-api.h5p.org/v1/metadata';
    public librariesUrl: string = '/libraries';
    public libraryConfig: { [machineName: string]: any };
    public libraryWhitelist: string = 'js css';
    public lrsContentTypes: string[] = [
        'H5P.Questionnaire',
        'H5P.FreeTextQuestion'
    ];
    public maxFileSize: number = 16 * 1024 * 1024;
    public maxTotalSize: number = 64 * 1024 * 1024;
    public paramsUrl: string = '/params';
    public platformName: string = 'H5P-Editor-NodeJs';
    public platformVersion: string = '0.10';
    public playerAddons?: {
        [machineName: string]: string[];
    };
    public playUrl: string = '/play';
    public proxy?: {
        host: string;
        port: number;
        protocol?: 'http' | 'https';
    };
    public sendUsageStatistics: boolean = false;
    public setFinishedUrl: string = '/finishedData';
    public setFinishedEnabled: boolean = true;
    public siteType: 'local' | 'network' | 'internet' = 'local';
    public contentUserStateSaveInterval: number | false = 5 * 1000; // the interval to save the contentUserData in milliseconds
    public temporaryFileLifetime: number = 120 * 60 * 1000; // 120 minutes
    public temporaryFilesUrl: string = '/temp-files';
    public uuid: string = '';

    private storage: IKeyValueStorage;

    /**
     * Loads all changeable settings from storage. (Should be called when the system initializes.)
     */
    public async load(): Promise<H5PConfig> {
        await this.loadSettingFromStorage('baseUrl');
        await this.loadSettingFromStorage('contentHubEnabled');
        await this.loadSettingFromStorage('contentHubMetadataRefreshInterval');
        await this.loadSettingFromStorage('contentTypeCacheRefreshInterval');
        await this.loadSettingFromStorage('contentUserStateSaveInterval');
        await this.loadSettingFromStorage('contentWhitelist');
        await this.loadSettingFromStorage('customization');
        await this.loadSettingFromStorage('disableFullscreen');
        await this.loadSettingFromStorage('editorAddons');
        await this.loadSettingFromStorage('enableLrsContentTypes');
        await this.loadSettingFromStorage('exportMaxContentPathLength');
        await this.loadSettingFromStorage('fetchingDisabled');
        await this.loadSettingFromStorage('hubContentTypesEndpoint');
        await this.loadSettingFromStorage('hubRegistrationEndpoint');
        await this.loadSettingFromStorage('libraryConfig');
        await this.loadSettingFromStorage('libraryWhitelist');
        await this.loadSettingFromStorage('maxFileSize');
        await this.loadSettingFromStorage('maxTotalSize');
        await this.loadSettingFromStorage('playerAddons');
        await this.loadSettingFromStorage('proxy');
        await this.loadSettingFromStorage('sendUsageStatistics');
        await this.loadSettingFromStorage('setFinishedEnabled');
        await this.loadSettingFromStorage('siteType');
        await this.loadSettingFromStorage('uuid');
        return this;
    }

    /**
     * Saves all changeable settings to storage. (Should be called when a setting was changed.)
     */
    public async save(): Promise<void> {
        await this.saveSettingToStorage('contentHubEnabled');
        await this.saveSettingToStorage('contentHubMetadataRefreshInterval');
        await this.saveSettingToStorage('contentTypeCacheRefreshInterval');
        await this.saveSettingToStorage('contentUserStateSaveInterval');
        await this.saveSettingToStorage('contentWhitelist');
        await this.saveSettingToStorage('customization');
        await this.saveSettingToStorage('disableFullscreen');
        await this.saveSettingToStorage('editorAddons');
        await this.saveSettingToStorage('enableLrsContentTypes');
        await this.saveSettingToStorage('exportMaxContentPathLength');
        await this.saveSettingToStorage('fetchingDisabled');
        await this.saveSettingToStorage('hubContentTypesEndpoint');
        await this.saveSettingToStorage('hubRegistrationEndpoint');
        await this.saveSettingToStorage('libraryConfig');
        await this.saveSettingToStorage('libraryWhitelist');
        await this.saveSettingToStorage('maxFileSize');
        await this.saveSettingToStorage('maxTotalSize');
        await this.saveSettingToStorage('playerAddons');
        await this.saveSettingToStorage('proxy');
        await this.saveSettingToStorage('sendUsageStatistics');
        await this.saveSettingToStorage('setFinishedEnabled');
        await this.saveSettingToStorage('siteType');
        await this.saveSettingToStorage('uuid');
    }

    /**
     * Loads a settings from the storage interface. Uses the default value
     * configured in this file if there is none in the configuration.
     * @param settingName
     * @returns the value of the setting
     */
    private async loadSettingFromStorage(settingName: string): Promise<any> {
        this[settingName] =
            (await this.storage?.load(settingName)) ?? this[settingName];
        return this[settingName];
    }

    /**
     * Saves a setting to the storage interface.
     * @param settingName
     */
    private async saveSettingToStorage(settingName: string): Promise<void> {
        await this.storage?.save(settingName, this[settingName]);
    }
}
