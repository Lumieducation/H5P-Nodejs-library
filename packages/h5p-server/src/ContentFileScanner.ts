import { ContentScanner } from './ContentScanner';
import Logger from './helpers/Logger';
import LibraryManager from './LibraryManager';
import { ILibraryName, ISemanticsEntry } from './types';

const log = new Logger('ContentFileScanner');

/**
 * Describes a reference to a file that is embedded inside the content.
 */
export interface IFileReference {
    /**
     * Information about where the file reference was found
     */
    context: {
        /**
         * The path of the object **inside the params tree**. You can use this
         * path to later modify the params, if needed.
         */
        jsonPath: string;
        /**
         * The raw parameters of the object.
         */
        params: any;
        /**
         * The semantic structure of the object (as defined in semantics.json).
         *
         * Can be null or undefined if there is no semantic structure for the
         * element (happens if the file can be found somewhere that is not
         * described by the semantics)
         */
        semantics?: ISemanticsEntry;
    };
    /**
     * The path of the file **inside the H5P content directory** as can be used
     * to reference it in ContentManager (without any suffixes like #tmp).
     */
    filePath: string;
    /**
     * The mime type specified in the params
     */
    mimeType?: string;
    /**
     * If true, the file was marked as temporary (by the #tmp suffix). The
     * suffix is **not included** in filePath
     */
    temporary: boolean;
}

/**
 * Scans the content parameters (=content.json) of a piece of content and
 * returns a list of references to file that are embedded inside the content.
 */
export class ContentFileScanner extends ContentScanner {
    constructor(libraryManager: LibraryManager) {
        super(libraryManager);
        log.info('initialize');
    }

    /**
     * Used to differentiate between local files and URLs.
     */
    private static urlRegExp: RegExp = /^https?:\/\//;

    /**
     * Loads the specified content from the ContentManager and scans its
     * parameters (= content.json) for references to local files (= audio,
     * video, images, generic files).
     * @param contentId the content to scan
     * @param user the user who wants to access the file
     * @returns a list of local files
     */
    public async scanForFiles(
        mainParams: any,
        mainLibraryName: ILibraryName
    ): Promise<IFileReference[]> {
        const results: IFileReference[] = [];
        await this.scanContent(
            mainParams,
            mainLibraryName,
            (semantics, params, jsonPath) => {
                log.debug(
                    `Checking entry ${jsonPath} (type ${semantics.type})`
                );
                switch (semantics.type) {
                    case 'file':
                    case 'image': {
                        log.debug(`found ${semantics.type} element`);
                        const element = this.pushIfDefined(
                            results,
                            this.checkFileElement(semantics, params, jsonPath)
                        );
                        if (element) {
                            log.debug(
                                `Found file is a reference to ${element.filePath}`
                            );
                        }
                        if (params.originalImage) {
                            const originalImageElement = this.pushIfDefined(
                                results,
                                this.checkFileElement(
                                    null,
                                    params.originalImage,
                                    `${jsonPath}.originalImage`
                                )
                            );
                            if (originalImageElement) {
                                log.debug(
                                    `Found file is a reference to ${originalImageElement.filePath}`
                                );
                            }
                        }
                        return true; // returning true aborts further recursion
                    }
                    case 'video':
                    case 'audio':
                        if (Array.isArray(params)) {
                            for (
                                let index = 0;
                                index < params.length;
                                index += 1
                            ) {
                                const arrayElement = this.pushIfDefined(
                                    results,
                                    this.checkFileElement(
                                        null,
                                        params[index],
                                        `${jsonPath}[${index}]`
                                    )
                                );
                                if (arrayElement) {
                                    log.debug(
                                        `Found file is a reference to ${arrayElement.filePath}`
                                    );
                                }
                            }
                        }
                        return true; // returning true aborts further recursion
                    default:
                        return false;
                }
            }
        );
        return results;
    }

    /**
     * Checks if an element in the parameter tree contains a valid reference to
     * a local file and removes temporary markers.
     * @param semantics The semantic structure of the element to check
     * @param params the parameter object of the element to check
     * @param jsonPath the JSONPath at which the element can be found in the
     * parameter object
     * @returns an object with information about the file reference; undefined
     * if the file reference is invalid
     */
    private checkFileElement(
        semantics: ISemanticsEntry,
        params: any,
        jsonPath: string
    ): IFileReference {
        if (!params.path) {
            // Path shouldn't be empty, but we simply ignore the entry in this
            // case.
            return undefined;
        }
        if (ContentFileScanner.urlRegExp.test(params.path)) {
            // If the file is a reference to a URL, we don't return it.
            return undefined;
        }

        let temporary = false;
        let cleanFileReferencePath = params.path;
        if (params.path.endsWith('#tmp')) {
            // files marked as temporary will be identified as such
            temporary = true;
            cleanFileReferencePath = params.path.substr(
                0,
                params.path.length - 4
            );
        }

        return {
            context: { semantics, params, jsonPath },
            filePath: cleanFileReferencePath,
            mimeType: params.mime,
            temporary
        };
    }

    /**
     * Helper function that pushes an item to an array if the item is defined.
     * @param array the array to push to
     * @param item the item to push
     * @returns the item (if defined); otherwise undefined
     */
    private pushIfDefined<T>(array: T[], item: T): T {
        if (item !== undefined) {
            array.push(item);
            return item;
        }
        return undefined;
    }
}
