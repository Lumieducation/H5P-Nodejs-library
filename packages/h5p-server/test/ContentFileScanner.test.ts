import * as fsExtra from 'fs-extra';
import jsonpath from 'jsonpath';
import * as path from 'path';
import { withDir } from 'tmp-promise';

import { ContentFileScanner } from '../src/ContentFileScanner';
import ContentManager from '../src/ContentManager';
import H5PConfig from '../src/implementation/H5PConfig';
import FileContentStorage from '../src/implementation/fs/FileContentStorage';
import FileLibraryStorage from '../src/implementation/fs/FileLibraryStorage';
import LibraryManager from '../src/LibraryManager';
import PackageImporter from '../src/PackageImporter';
import { ContentId, IUser } from '../src/types';
import ContentStorer from '../src/ContentStorer';

import User from './User';
import { getContentDetails } from './ContentScanner.test';

describe('ContentFileScanner', () => {
    async function createContentFileScanner(
        file: string,
        user: IUser,
        tmpDirPath: string
    ): Promise<{
        contentId: ContentId;
        contentManager: ContentManager;
        contentScanner: ContentFileScanner;
    }> {
        // create required dependencies
        const contentDir = path.join(tmpDirPath, 'content');
        const libraryDir = path.join(tmpDirPath, 'libraries');
        await fsExtra.ensureDir(contentDir);
        await fsExtra.ensureDir(libraryDir);

        const contentManager = new ContentManager(
            new FileContentStorage(contentDir)
        );
        const libraryManager = new LibraryManager(
            new FileLibraryStorage(libraryDir)
        );

        // install content & libraries
        const packageImporter = new PackageImporter(
            libraryManager,
            new H5PConfig(null),
            contentManager,
            new ContentStorer(contentManager, libraryManager, undefined)
        );
        const contentId = (
            await packageImporter.addPackageLibrariesAndContent(file, user)
        ).id;

        // create ContentScanner
        return {
            contentId,
            contentManager,
            contentScanner: new ContentFileScanner(libraryManager)
        };
    }

    it('finds the image in H5P.Blanks example', async () => {
        await withDir(
            async ({ path: tmpDirPath }) => {
                const user = new User();
                user.canUpdateAndInstallLibraries = true;

                const {
                    contentScanner,
                    contentId,
                    contentManager
                } = await createContentFileScanner(
                    path.resolve('test/data/hub-content/H5P.Blanks.h5p'),
                    user,
                    tmpDirPath
                );

                const { params, mainLibraryName } = await getContentDetails(
                    contentId,
                    user,
                    contentManager
                );

                const foundImages = await contentScanner.scanForFiles(
                    params,
                    mainLibraryName
                );

                expect(foundImages.length).toEqual(1);
                expect(path.normalize(foundImages[0].filePath)).toEqual(
                    path.normalize('images/file-5885c18261805.jpg')
                );
                const parameters = await contentManager.getContentParameters(
                    contentId,
                    user
                );
                expect(
                    jsonpath.query(
                        parameters,
                        foundImages[0].context.jsonPath
                    )[0].path
                ).toEqual(foundImages[0].filePath);
            },
            { keep: false, unsafeCleanup: true }
        );
    });

    it('finds the image in H5P.Questionnaire example (= weird group single fields entry behaviour)', async () => {
        await withDir(
            async ({ path: tmpDirPath }) => {
                const user = new User();
                user.canUpdateAndInstallLibraries = true;

                const {
                    contentScanner,
                    contentId,
                    contentManager
                } = await createContentFileScanner(
                    path.resolve('test/data/hub-content/H5P.Questionnaire.h5p'),
                    user,
                    tmpDirPath
                );

                const { params, mainLibraryName } = await getContentDetails(
                    contentId,
                    user,
                    contentManager
                );

                const foundImages = await contentScanner.scanForFiles(
                    params,
                    mainLibraryName
                );

                expect(foundImages.length).toEqual(1);
                expect(path.normalize(foundImages[0].filePath)).toEqual(
                    path.normalize('images/file-5a4d06a8cbabc.jpg')
                );
            },
            { keep: false, unsafeCleanup: true }
        );
    });
});
