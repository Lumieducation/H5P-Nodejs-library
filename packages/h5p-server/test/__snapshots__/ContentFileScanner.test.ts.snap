// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentScanner aborts scanning when requested 1`] = `
[
  "$.a11yCheck",
  "$.a11yCheckingModeHeader",
  "$.a11yRetry",
  "$.a11yShowSolution",
  "$.answerIsCorrect",
  "$.answerIsWrong",
  "$.answeredCorrectly",
  "$.answeredIncorrectly",
  "$.behaviour",
  "$.checkAnswer",
  "$.confirmCheck",
  "$.confirmRetry",
  "$.confirmRetry.body",
  "$.confirmRetry.cancelLabel",
  "$.confirmRetry.confirmLabel",
  "$.confirmRetry.header",
  "$.inputHasTipLabel",
  "$.inputLabel",
  "$.media",
  "$.notFilledOut",
  "$.overallFeedback",
  "$.overallFeedback[0]",
  "$.overallFeedback[0].feedback",
  "$.overallFeedback[0].from",
  "$.overallFeedback[0].to",
  "$.questions",
  "$.questions[0]",
  "$.questions[1]",
  "$.questions[2]",
  "$.scoreBarLabel",
  "$.showSolutions",
  "$.solutionLabel",
  "$.submitAnswer",
  "$.text",
  "$.tipLabel",
  "$.tryAgain",
]
`;

exports[`ContentScanner scans the semantic structure of H5P.Blanks example 1`] = `
[
  "$.a11yCheck",
  "$.a11yCheckingModeHeader",
  "$.a11yRetry",
  "$.a11yShowSolution",
  "$.answerIsCorrect",
  "$.answerIsWrong",
  "$.answeredCorrectly",
  "$.answeredIncorrectly",
  "$.behaviour",
  "$.behaviour.acceptSpellingErrors",
  "$.behaviour.autoCheck",
  "$.behaviour.caseSensitive",
  "$.behaviour.confirmCheckDialog",
  "$.behaviour.confirmRetryDialog",
  "$.behaviour.enableCheckButton",
  "$.behaviour.enableRetry",
  "$.behaviour.enableSolutionsButton",
  "$.behaviour.separateLines",
  "$.behaviour.showSolutionsRequiresInput",
  "$.checkAnswer",
  "$.confirmCheck",
  "$.confirmCheck.body",
  "$.confirmCheck.cancelLabel",
  "$.confirmCheck.confirmLabel",
  "$.confirmCheck.header",
  "$.confirmRetry",
  "$.confirmRetry.body",
  "$.confirmRetry.cancelLabel",
  "$.confirmRetry.confirmLabel",
  "$.confirmRetry.header",
  "$.inputHasTipLabel",
  "$.inputLabel",
  "$.media",
  "$.media.disableImageZooming",
  "$.media.type",
  "$.media.type.params.alt",
  "$.media.type.params.contentName",
  "$.media.type.params.decorative",
  "$.media.type.params.expandImage",
  "$.media.type.params.file",
  "$.media.type.params.minimizeImage",
  "$.media.type.params.title",
  "$.notFilledOut",
  "$.overallFeedback",
  "$.overallFeedback[0]",
  "$.overallFeedback[0].feedback",
  "$.overallFeedback[0].from",
  "$.overallFeedback[0].to",
  "$.questions",
  "$.questions[0]",
  "$.questions[1]",
  "$.questions[2]",
  "$.scoreBarLabel",
  "$.showSolutions",
  "$.solutionLabel",
  "$.submitAnswer",
  "$.text",
  "$.tipLabel",
  "$.tryAgain",
]
`;
